### PR TITLE
Remove superfluous DISTINCT

### PR DESCRIPTION
**Pull request by dey4ss**
Thanks to @Bouncner, I read the following note in [a paper](https://www.vldb.org/pvldb/vol16/p1208-liu.pdf):
> It took more than two years for PostgreSQL developers to implement a pattern that
removes the DISTINCT clause if the result is unique by definition, and this feature is
yet to be merged [26].

That's true and it does not look like they will even merge it soon: https://commitfest.postgresql.org/35/2433/

Thus, I was triggered to implement it in Hyrise and prove we can do this in less than half an hour. It does not change anything in benchmarks, but now, we have it.
(To be fair: The Postgres contributors had to implement passing UCCs in the LQP and tried multiple versions. Also, I needed more than half an hour with all tests and an edge case I forgot to cover.)

### CODE REVIEW

**Comment by Bouncner** [comment]
> Thus, I was triggered to implement it in Hyrise and prove we can do this in less than half an hour.

Angeber.

**Review by Bouncner — COMMENTED** [review]
Only minor things.

File: src/lib/sql/sql_translator.cpp [hunk 1]
```diff
@@ -1186,11 +1186,11 @@ void SQLTranslator::_translate_select_groupby_having(const hsql::SelectStatement
   //
   // This might create unnecessary aggregate nodes when we already have an aggregation that creates unique results:
   // `SELECT DISTINCT a, MIN(b) FROM t GROUP BY a` would have one aggregate that groups by a and calculates MIN(b), and
-  // one that groups by both a and MIN(b) without calculating anything. Fixing this should be done by an optimizer rule
-  // that checks for each GROUP BY whether it guarantees the results to be unique or not. Doable, but no priority.
+  // one that groups by both a and MIN(b) without calculating anything. Fixing this is done by an optimizer rule
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/lib/sql/sql_translator.cpp:1189]: I do. We did two aggregates before, now, we do one (the table is unique for the group keys after the first aggregate).

File: src/test/lib/optimizer/strategy/dependent_group_by_reduction_rule_test.cpp [hunk 2]
```diff
@@ -141,8 +144,8 @@ TEST_F(DependentGroupByReductionRuleTest, IncompleteKey) {
     stored_table_node_b);
   // clang-format on
 
-  const auto actual_lqp = apply_rule(rule, lqp);
   const auto expected_lqp = lqp->deep_copy();
+  const auto actual_lqp = apply_rule(rule, lqp);
```
[the remark this answers is not in the document]
Comment (dey4ss) [reply src/test/lib/optimizer/strategy/dependent_group_by_reduction_rule_test.cpp:148]: As discussed offline, we copied the optimized version before, which is of course equal, but that's not what is intended.

File: src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp [hunk 3]
```diff
@@ -102,7 +106,9 @@ void DependentGroupByReductionRule::_apply_to_plan_without_subqueries(
         return LQPVisitation::VisitInputs;
       }
 
-      // Else, replace it with a ProjectionNode.
+      // Else, replace it with a ProjectionNode. For instance, SELECT DISTINCT n_name, n_regionkey FROM nation (the
+      // column order is different than in the original table) turns from [ Aggregate GROUP BY n_name, n_regionkey ] to
```
Comment (Bouncner) [inline src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp:110]: "turns from to"? Sounds strange.

**Review by Bouncner — APPROVED** [review]
Just a minor note on a strange sounding sentencing. Otherwise ready to zünd.

File: src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp (hunk 3 above)
Comment (dey4ss) [reply src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp:110]: `, which we represent by [ Aggregate GROUP BY n_name, n_regionkey ], is rewritten to ...`?

Comment (Bouncner) [reply src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp:110]: Jop

Comment (Bouncner) [reply src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp:110]: Hmm ... maybe I was drunk? It doesn't sound problematic to me today. Feel free to ignore. I might have misread the sentence. :-/

Comment (dey4ss) [reply src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp:110]: 🤣 
🧨 then?
